test/lsp.py

```diff
@@ -180,15 +282,411 @@ class Counter:
+                        )
+
+                    self.suite.expect_diagnostic(
+                        diagnostic,
+                        code=expected_diagnostic.code,
+                        marker=lMarkers[expected_diagnostic.marker]
+                    )
+
+        except Exception as e:
+            print(e)
+            self.close_all_open_files()
+            raise
+
+    def close_all_open_files(self):
+        for test in self.open_tests:
+            self.solc.send_message(
+                'textDocument/didClose',
+                { 'textDocument': { 'uri': self.suite.get_test_file_uri(test) }}
+            )
+            self.suite.wait_for_diagnostics(self.solc)
+
+        self.open_tests.clear()
+
+    def test_methods(self) -> bool:
+        """
+        Test all methods. Returns False if a reparsing is required, else True
+        """
+        try:
+            # Now handle each request/response pair in the test definition
+            for testcase in self.parsed_testcases:
+                try:
+                    self.run_testcase(testcase)
+                except JSONExpectationFailed as e:
+                    result = self.user_interaction_failed_method_test(testcase, e.actual, e.expected)
+
+                    if result == self.TestResult.Reparse:
+                        return False
+
+            return True
+        except TestParserException as e:
+            print(e)
+            print(e.result)
+            raise
+        except Exception as e:
+            print(e)
+            raise
+        finally:
+            self.close_all_open_files()
+
+    def user_interaction_failed_method_test(self, testcase, actual, expected):
+        actual_pretty = self.suite.replace_ranges_with_tags(actual)
+
+        if expected is None:
+            print("Failed to parse expected response, received: \n" + actual)
+        else:
+            print("Expected:\n" + \
+                self.suite.replace_ranges_with_tags(expected) + \
+                "\nbut got:\n" + actual_pretty
+            )
+
+        while True:
+            print("(u)pdate/(r)etry/(i)gnore?")
+            user_response = sys.stdin.read(1)
+            if user_response == "i":
+                return self.TestResult.SuccessOrIgnored
+
+            if user_response == "u":
+                actual = actual["result"]
+                self.content = self.content[:testcase.responseBegin] + \
+                    prependComments("<- " + self.suite.replace_ranges_with_tags(actual)) + \
+                    self.content[testcase.responseEnd:]
+
+                with open(self.suite.get_test_file_path(self.test_name), mode="w", encoding="utf-8", newline='') as f:
+                    f.write(self.content)
+                return self.TestResult.Reparse
+            if user_response == "r":
+                return self.TestResult.Reparse
+
+            print("Invalid response.")
+
+
+    def run_testcase(self, testcase):
```

Comment:
```suggestion
    def run_testcase(self, testcase: Tuple[str, str]):
```
or similar - is it possible to specify the type? Otherwise it should be documented at least.